runtime/json-writer.cpp

```diff
@@ -0,0 +1,215 @@
+  if (!RegisterValue()) {
+    return false;
+  }
+  static_SB.append("null", 4);
+  return true;
+}
+
+bool JsonWriter::Key(std::string_view key) noexcept {
+  if (stack_top_ == -1 || stack_[stack_top_].in_array) {
+    error_.append("json key is allowed only inside object");
+    return false;
+  }
+  if (stack_[stack_top_].values_count) {
+    static_SB << ',';
+  }
+  writeChar('\n');
+  writeIndent();
+  static_SB << '"';
+  static_SB.append(key.data(), key.size());
+  static_SB << '"';
+  static_SB << ':';
+  writeChar(' ');
+  return true;
+}
+
+bool JsonWriter::StartObject() noexcept {
+  return newLevel(false);
+}
+
+bool JsonWriter::EndObject() noexcept {
+  return exitLevel(false);
+}
+
+bool JsonWriter::StartArray() noexcept {
+  return newLevel(true);
+}
+
+bool JsonWriter::EndArray() noexcept {
+  return exitLevel(true);
+}
+
+bool JsonWriter::IsComplete() const noexcept {
+  return error_.empty() && stack_top_ == -1 && has_root_;
+}
+
+string JsonWriter::GetError() const noexcept {
+  return error_;
+}
+
+string JsonWriter::GetJson() const noexcept {
+  return static_SB.str();
+}
+
+bool JsonWriter::newLevel(bool is_array) noexcept {
+  if (!RegisterValue()) {
+    return false;
+  }
+  ++stack_top_;
+  if (stack_top_ == MAX_DEPTH) {
+    error_.append("stack overflow, max depth level is ");
+    error_.append(static_cast<std::int64_t>(MAX_DEPTH));
+    return false;
+  }
+  stack_[stack_top_] = NestedLevel{.in_array = is_array};
+
+  static_SB << (is_array ? '[' : '{');
+  setIndent();
+  return true;
+}
+
+bool JsonWriter::exitLevel(bool is_array) noexcept {
+  if (stack_top_ == -1) {
+    error_.append("brace disbalance");
+    return false;
+  }
+  auto cur_level = stack_[stack_top_];
+  --stack_top_;
+  if (cur_level.in_array != is_array) {
+    error_.append("attempt to enclosure ");
+    error_.push_back(cur_level.in_array ? '[' : '{');
+    error_.append(" with ");
+    error_.push_back(is_array ? ']' : '}');
+    return false;
+  }
+
+  if (cur_level.values_count) {
+     writeChar('\n');
+     endIndent();
+     writeIndent();
+  } else {
+    endIndent();
+  }
+  static_SB << (is_array ? ']' : '}');
+  return true;
+}
+
+bool JsonWriter::RegisterValue() noexcept {
+  if (has_root_ && stack_top_ == -1) {
+    error_.append("attempt to set value twice in a root of json");
+    return false;
+  }
+  if (stack_top_ == -1) {
+    has_root_ = true;
+    return true;
+  }
+  if (stack_[stack_top_].in_array) {
+    if (stack_[stack_top_].values_count) {
+      static_SB << ',';
+    }
+    writeChar('\n');
+    writeIndent();
+  }
+  ++stack_[stack_top_].values_count;
+
+  return true;
+}
+
+void JsonWriter::setIndent() noexcept {
+  indent_ += 4;
+}
+
+void JsonWriter::endIndent() noexcept {
+  indent_ -= 4;
+}
+
+void JsonWriter::writeIndent() noexcept {
+  if (pretty_print_ && indent_) {
+    static_SB.reserve(indent_);
+    for (std::size_t i = 0; i < indent_; ++i) {
+      static_SB.append_char(' ');
```

Comment:
Why do you use `append_char` here, but `<<` in a method below?

Reply:
Because `<<` reserve space for 1 char, but here I want to write 4 chars, so reserve required amount by calling `reserve(4)`